### ISSUE TITLE
[Bug] My software stucks on browser.close with plugin-stealth 2.7.x version

### ISSUE THREAD

**Issue by Dam998** [berstend/puppeteer-extra#421]
My software stucks on browser.close with last version (puppeteer-extra-plugin-stealth)

I'm using more than 1 browser open at the same time, and when I'm trying to close 1 my software stucks.
this has never happened with version 2.6.7 or before

my code:

```javascript
var pages = await this.browser.pages()
for (let i = 0; i < pages.length; i++) {
    await pages[i].close();
}
await this.browser.close();
```

Tested with 2.7.x version

if i downgrade to 2.6.7 everything works fine

**Comment by berstend** [comment]
How are you opening the browsers, `.launch` or `.connect`?

**Comment by Dam998** [comment]
.launch

**Comment by berstend** [comment]
Could you remove the `user-agent-override` evasion and test if it works then?

```js
const pluginStealth = require('puppeteer-extra-plugin-stealth')()
pluginStealth.enabledEvasions.delete('user-agent-override')
puppeteer.use(pluginStealth)
```

**Comment by berstend** [comment]
The issue might be related to https://github.com/berstend/puppeteer-extra/pull/413 - we save some state there and added functionality depending on how the browser is launched

**Comment by Dam998** [comment]
if I remove user-agent-override is working

can I ask you why? if this is correct and nothing change at the end
or if you suggest me to launch browser in a different way

**Comment by berstend** [comment]
So it works without the `user-agent-override` evasion?

The reason is to narrow down the problem, we made a larger update to the evasion so it makes sense the issue is somewhere there.

If removing `user-agent-override` fixes the issue in 2.7.x then you just need to wait for us to develop a fix, until then you need to use the older version for your use-case.

**Comment by berstend** [comment]
@Dam998 did you mean opening multiple browsers or multiple pages in a single browser?

Your words indicate the former, your code the latter

**Comment by Dam998** [comment]
opening multiple browsers (more browsers, each browser with 2 pages)

I created a class that on start() open a browser and then close it at the end, this is why

**Comment by Niek** [comment]
@Dam998 do you also mind sharing a minimal reproducible code sample so we can check?

**Comment by berstend** [comment]
Are you able to make a small self-contained test file? this helps us reproduce the issue quickly. Thanks!

**Comment by Dam998** [comment]
I'll try, give me some minutes

**Comment by Dam998** [comment]
I reproduced it with this simple script:

```javascript
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
const locateChrome = require('locate-chrome');
const delay = require('delay');

(async () => {
    console.log('start')
    for (var i = 0; i < 20; i++) {
        (async () => {
            console.log('start')
            try {
                var chromeOptions = {
                    headless: false,
                    ignoreDefaultArgs: ['--disable-extensions'],
                    ignoreHTTPSErrors: true,
                    args: [
                        '--disable-web-security',
                        '--disable-dev-shm-usage',
                        '--no-sandbox',
                        '--disable-setuid-sandbox',
                        '--disable-gpu',
                        '--ignore-certificate-errors',
                        '--enable-features=NetworkService'
                    ],
                };

                let chromePath = await locateChrome()
                chromeOptions.executablePath = chromePath;

                puppeteer.use(StealthPlugin())
                const browser = await puppeteer.launch(chromeOptions);
                const page = await browser.newPage();
                
                await page.goto('https://github.com/berstend/puppeteer-extra/issues/421', { 'waitUntil': 'domcontentloaded', timeout: 60000 });
                await delay(13000);

                var pages = await browser.pages()
                for (let j = 0; j < pages.length; j++) {
                    await pages[j].close();
                }
                await browser.close();
            }
            catch (err) {
                console.log(err)
            }
        })();
        await delay(10000);
    }
})();
```

this script just open more browsers and then close them continusly
with 2.6.7 version everything works good, with 2.7.x after the first browser close, the script freezes

and with 2.7.x if you try to do CTRL + C in terminal it not wll close, with 2.6.7 yes

**Comment by berstend** [comment]
Thanks! can't test right now but does it make a difference if you move `puppeteer.use(StealthPlugin())` outside of the loop? It's not necessary to do that 20 times, once is enough. :-)

**Comment by Dam998** [comment]
nothing change

this is only an example, I made this little mistake here but nothing change if I move it outside of the loop

**Comment by berstend** [comment]
Gotcha. Thanks for the test case, we'll look into it. :-)

**Comment by Niek** [comment]
I cannot reproduce this issue, your test script works fine for me @Dam998. I only got some memleak warnings but that's to be expected:
```bash
$ node test.js
start
start
start
start
start
start
start
start
start
start
(node:70151) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 exit listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
(node:70151) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGINT listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(node:70151) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGTERM listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(node:70151) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGHUP listeners added to [process]. Use emitter.setMaxListeners() to increase limit
start
start
start
start
start
start
start
start
start
start
start
```

**Comment by Dam998** [comment]
@Niek 

![image](https://user-images.githubusercontent.com/23696659/108425893-677fd580-723b-11eb-96b4-62a0c089532d.png)

**Comment by Dam998** [comment]
I tested again my script with 2.7.4 version and I'm getting same error
with 2.6.7 works fine

**Comment by berstend** [comment]
@Dam998 pretty sure this is related to you using Windows in one way or another - our core team is all on POSIX.

Are you using WSL to run this stuff?

**Comment by berstend** [comment]
The issue is most likely somewhere in the user-data-dir plugin:
https://github.com/berstend/puppeteer-extra/blob/master/packages/puppeteer-extra-plugin-user-data-dir/index.js

Can you run your script with `DEBUG=puppeteer-extra-plugin:user-data-dir myscript.js`?

**Comment by Dam998** [comment]
with version 2.7.4:
```
$ DEBUG=puppeteer-extra-plugin:user-data-dir node testBug.js

start
start
  puppeteer-extra-plugin:user-data-dir initialized {
  deleteTemporary: true,
  deleteExisting: false,
  files: [],
  folderPath: 'C:\\Users\\dpapi\\AppData\\Local\\Temp',
  folderPrefix: 'puppeteer_dev_profile-'
} +0ms
  puppeteer-extra-plugin:user-data-dir created custom dir C:\Users\dpapi\AppData\Local\Temp\puppeteer_dev_profile-sCe9ey +5ms
  puppeteer-extra-plugin:user-data-dir Wrote file C:\Users\dpapi\AppData\Local\Temp\puppeteer_dev_profile-sCe9ey\Default\Preferences +12ms
start
  puppeteer-extra-plugin:user-data-dir created custom dir C:\Users\dpapi\AppData\Local\Temp\puppeteer_dev_profile-O8kLVd +10s
  puppeteer-extra-plugin:user-data-dir Wrote file C:\Users\dpapi\AppData\Local\Temp\puppeteer_dev_profile-O8kLVd\Default\Preferences +1ms
  puppeteer-extra-plugin:user-data-dir onClose +5s
  puppeteer-extra-plugin:user-data-dir removeUserDataDir +1ms
```
  
  
  with version 2.7.4 everything stuck, the terminal too doesn't responde, I need always to force close it
  but only if I reproduce this situation, if I use only 1 browser at a time there are no problems

**Comment by Niek** [comment]
@berstend what do you think about bumping `fs-extra` to `@9`? See also: https://github.com/jprichardson/node-fs-extra/blob/master/CHANGELOG.md

**Comment by berstend** [comment]
Yeah, bumping `fs-extra` might fix this issue

**Comment by Niek** [comment]
@Dam998 Please check with the latest master branch.

**Comment by berstend** [comment]
> @Dam998 Please check with the latest master branch.

Published as `puppeteer-extra-plugin-user-data-dir@2.2.12` :-)

**Comment by Dam998** [comment]
hello
I have updated stealth and puppeteer-extra to the last version (2.7.6, 3.1.18)
I installed puppeteer-extra-plugin-user-data-dir@2.2.12 and started my script but I'm getting always the same error, it seems not fixed

maybe I didn't understand how I have to use this new plugin (puppeteer-extra-plugin-user-data-dir) because I have ever used it bf

**Comment by alamothe** [comment]
Hello, just wanted to report the same problem. 

- The problem started when we upgraded `2.6.5` to `2.7.6`. 
- I did not observe the issue on my MacOS (this is where I'm developing).
- In production we use Windows, and it hangs on close.
- We open multiple browsers. We use `launch` and real Chrome 89.0.4389.90.
- Reverting back to `2.6.5` fixes the problem.

**Comment by Niek** [comment]
@alamothe would it be possible to try to test versions `2.6.6`, `2.6.7` and earlier `2.7.x` versions to see where it breaks exactly?

**Comment by Dam998** [comment]
@Niek @berstend do you have any news about this issue?

**Comment by Dam998** [comment]
the error happen when puppetter try to close 1 browser
then everything ll stuck
i tested with chrome and chromium

if I open only browsers everything work perfectly but when I close only 1 everything stuck

**Comment by dalespiteri** [comment]
This happens for me as well with the default run settings listed in the documentation. `browser.close` locks up my terminal process. If I remove the `puppeteer-extra` requirement, and just go with the the vanilla `puppeteer`, everything is fine. There's definitely a core issue with this package. 

Literally just running the quickstart settings doesn't work:

```
// puppeteer-extra is a drop-in replacement for puppeteer,
// it augments the installed puppeteer with plugin functionality.
// Any number of plugins can be added through `puppeteer.use()`
const puppeteer = require('puppeteer-extra')

// Add stealth plugin and use defaults (all tricks to hide puppeteer usage)
const StealthPlugin = require('puppeteer-extra-plugin-stealth')
puppeteer.use(StealthPlugin())

// Add adblocker plugin to block all ads and trackers (saves bandwidth)
const AdblockerPlugin = require('puppeteer-extra-plugin-adblocker')
puppeteer.use(AdblockerPlugin({ blockTrackers: true }))

// That's it, the rest is puppeteer usage as normal 😊
puppeteer.launch({ headless: true }).then(async browser => {
  const page = await browser.newPage()
  await page.setViewport({ width: 800, height: 600 })

  console.log(`Testing adblocker plugin..`)
  await page.goto('https://www.vanityfair.com')
  await page.waitForTimeout(1000)
  await page.screenshot({ path: 'adblocker.png', fullPage: true })

  console.log(`Testing the stealth plugin..`)
  await page.goto('https://bot.sannysoft.com')
  await page.waitForTimeout(5000)
  await page.screenshot({ path: 'stealth.png', fullPage: true })

  console.log(`All done, check the screenshots. ✨`)
  await browser.close()
})
```

**Comment by mathias22osterhagen22** [comment]
Same for me, 
I'm using 3 browser instances, when the first one ".close" the two other seems to stop activity and my program soft lock as nothing resolve.

I added few lines in `Connection.js` of puppeteer to debug.
```js
 _onClose() {
        if (this._closed)
            return;
        this._closed = true;
        this._transport.onmessage = null;
        this._transport.onclose = null;
        for (const callback of this._callbacks.values())
            callback.reject(rewriteError(callback.error, `Protocol error (${callback.method}): Target closed.`));
        this._callbacks.clear();
        for (const session of this._sessions.values())
            session._onClosed();
        this._sessions.clear();
        console.log('before');
        this.emit(exports.ConnectionEmittedEvents.Disconnected);
        console.log('after');
    }
```

"after" never shows up on my console, I have no idea why.

**Comment by Niek** [comment]
@mathias22osterhagen22 are you using Windows or a other platform?

**Comment by mathias22osterhagen22** [comment]
@Niek, yes I'm using Windows 64 PRO, 10.0.19042 Build 19042, 32Giga ram, I7-10700 CPU @ 2.90Ghz

**Comment by dev-hyperweb** [comment]
@mathias22osterhagen22 

Can you guy confirm my solution are solves this ?
https://github.com/berstend/puppeteer-extra/pull/530

**Comment by k-funk** [comment]
I'm having this issue as well, but on MacOS:
```
  System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 14.12.0 - /usr/local/bin/node
    Yarn: Not Found
    npm: 6.14.8 - /usr/local/bin/npm
  Browsers:
    Chrome: 92.0.4515.107
    Edge: Not Found
    Firefox: 86.0
    Safari: 14.1.1

...

    "puppeteer": "^10.1.0",
    "puppeteer-extra": "^3.1.18",
    "puppeteer-extra-plugin-stealth": "^2.7.8"
```

This suggested code makes my SIGINTs work as expected again. Without the 2nd line, I have to ctrl-c 3 times, and even then, the process doesn't get killed.
```
const stealthPlugin = StealthPlugin()
stealthPlugin.enabledEvasions.delete('user-agent-override')
puppeteer.use(stealthPlugin)
```

You can test this out yourself [in this project](https://github.com/k-funk/us-cr-embassy-passport-appointment-scraper).

**Comment by KhalilMohammad** [comment]
I get this issue if I use ```puppeteer-extra``` with ```puppeteer-page-proxy```.
It freezes on ```browser.close()```

**Comment by alamothe** [comment]
@k-funk @KhalilMohammad this issue is Windows specific, please create a separate issue. Thanks!

**Comment by LinkTree3** [comment]
I am seeing the same issue right now.
I am on windows and when I add 'puppeteer-extra-plugin-stealth' and use `puppeteer.use(StealthPlugin())` I can't use ctrl+c to terminate the process. :(

Is there any solution for this?

**Comment by berstend** [comment]
Alright, to summarize what we know so far:
- The reason this happens in version 2.7 and up is that we started using the [puppeteer-extra-plugin-user-data-dir](https://github.com/berstend/puppeteer-extra/blob/master/packages/puppeteer-extra-plugin-user-data-dir) plugin in the `user-agent-override` evasion (unfortunately that's necessary to properly spoof locales)
- As to why `user-agent-override` evasion causes this issue is not entirely clear yet, but my gut tells it must be this code: https://github.com/berstend/puppeteer-extra/blob/0049d6010311505f27e7f3be804bb198e2c09aa2/packages/puppeteer-extra-plugin-user-data-dir/index.js#L111-L116

Can someone experiencing this issue monkey patch their `./node_modules/puppeteer-extra-plugin-user-data-dir/index.js` file and comment out this `onClose` handler to see if this fixes it?

**Comment by berstend** [comment]
more specifically the onClose handler will result in this sync call:
https://github.com/berstend/puppeteer-extra/blob/0049d6010311505f27e7f3be804bb198e2c09aa2/packages/puppeteer-extra-plugin-user-data-dir/index.js#L67-L76

**Comment by LinkTree3** [comment]
> Alright, to summarize what we know so far:
> 
> * The reason this happens in version 2.7 and up is that we started using the [puppeteer-extra-plugin-user-data-dir](https://github.com/berstend/puppeteer-extra/blob/master/packages/puppeteer-extra-plugin-user-data-dir) plugin in the `user-agent-override` evasion (unfortunately that's necessary to properly spoof locales)
> * As to why the `user-data-dir` plugin causes this behavior is not entirely clear yet, but my gut tells it must be this code: https://github.com/berstend/puppeteer-extra/blob/0049d6010311505f27e7f3be804bb198e2c09aa2/packages/puppeteer-extra-plugin-user-data-dir/index.js#L111-L116
> 
> Can someone experiencing this issue monkey patch their `./node_modules/puppeteer-extra-plugin-user-data-dir/index.js` file and comment out this `onClose` handler to see if this fixes it?

Hey thank you so much for looking into this! Yes I'll gladly check this and report. :)

**Comment by LinkTree3** [comment]
Yes your suggestion worked! :)

**Comment by berstend** [comment]
@LinkTree3 Cool, thanks for testing :) 

Would you be able to test if this change makes a difference (commenting the onClose handler back in)?
https://github.com/berstend/puppeteer-extra/pull/502/files

(basically replacing `rmdirSync` with `rmSync`)

**Comment by dev-hyperweb** [comment]
it clearly. this issued  stuck in onClose  event

so  if im not wrong, even  empty event  still stuck 
in now, I fixed with  onDisconnected  event  to use for my project

**Comment by LinkTree3** [comment]
> rmSync

This did not work.

**Comment by berstend** [comment]
> > rmSync
> 
> This did not work.

Ok, one last test :)

```js
fs.rm(this._userDataDir, { recursive: true, force: true }, console.log)

```

**Comment by LinkTree3** [comment]
I added a log right before the rmSync and when pressing ctrl+c it just keeps repeating the log

**Comment by berstend** [comment]
> YUP This did the trick! :)

Sweet! Thanks for testing 👍 So we know that (for whatever reason) removing the temporary profile folder synchronously can cause issues on windows, whereas doing it asynchronously works fine.

I'm gonna prepare a fix + new version within the next days, as time permits.

**Comment by LinkTree3** [comment]
> > YUP This did the trick! :)
> 
> Sweet! Thanks for testing 👍 So we know that (for whatever reason) removing the temporary profile folder synchronously can cause issues on windows, whereas doing it asynchronously works fine.
> 
> I'm gonna prepare a fix + new version within the next days, as time permits.

Thank you so much! looking forward to it :)

**Comment by LinkTree3** [comment]
berstend I am so sorry, but it seems that I might have had a mistake while testing it with `fs.rm(this._userDataDir, { recursive: true, force: true }, console.log)`

Now when I try it again it seems to not work :(

I am getting an EBUSY error and a bunch of "null" when I try to exit

[Error: EBUSY: resource busy or locked, unlink 'C:\Users\[username]\AppData\Local\Temp\puppeteer_dev_profile-1BHM2o\lockfile'] {
  errno: -4082,
  code: 'EBUSY',
  syscall: 'unlink',
  path: 'C:\\Users\\[username]\\AppData\\Local\\Temp\\puppeteer_dev_profile-1BHM2o\\lockfile'
}

**Comment by berstend** [comment]
@LinkTree3 no worries :) could you try if changing `onClose` to `onDisconnected` (with the original `fse.rmdirSync` code) does make a difference in your case?
https://github.com/berstend/puppeteer-extra/pull/530/files

**Comment by LinkTree3** [comment]
> @LinkTree3 no worries :) could you try if changing `onClose` to `onDisconnected` (with the original `fse.rmdirSync` code) does make a difference in your case?
> https://github.com/berstend/puppeteer-extra/pull/530/files
Yes this one seems to work! I tested it a couple times to make sure :)

**Comment by berstend** [comment]
@LinkTree3 awesome, we're gonna go with this fix then

**Comment by berstend** [comment]
This should be fixed now through #530 (thanks @dev-hyperweb 👍 )
```
Successfully published:
 - puppeteer-extra-plugin-stealth@2.7.9
 - puppeteer-extra-plugin-user-data-dir@2.2.13
 - puppeteer-extra-plugin-user-preferences@2.2.13
```
cc @LinkTree3 @Dam998

**Comment by berstend** [comment]
More properly fixed in #656
